lib/src/component/ref_util.dart

```diff
@@ -212,3 +214,282 @@ UiFactory<TProps> Function(UiFactory<TProps>) forwardRef<TProps extends UiProps>
+/// }
+/// ```
+///
+/// __Example 3:__ Forwarding refs in higher-order (all function) components
+///
+/// ```dart
+/// import 'dart:html';
+/// import 'package:over_react/over_react.dart';
+/// import 'package:over_react/react_dom.dart' as react_dom;
+/// import 'package:react/hooks.dart';
+///
+/// // ---------- Component Definitions ----------
+///
+/// final FancyButton = uiForwardRef<DomProps>((props, ref) {
+///     final classes = ClassNameBuilder.fromProps(props)..add('FancyButton');
+///
+///     return (Dom.button()
+///       ..addProps(getPropsToForward(props, onlyCopyDomProps: true))
+///       ..className = classes.toClassName()
+///       ..ref = ref
+///     )('Click me!');
+///   },
+///   Dom.button.asForwardRefConfig<DomProps>(displayName: 'FancyButton'),
+/// );
+///
+/// mixin LogPropsProps on UiProps {
+///   BuilderOnlyUiFactory<DomProps> builder;
+/// }
+///
+/// final LogProps = uiForwardRef<LogPropsProps>(
+///   (props, ref) {
+///     final prevPropsRef = useRef<LogPropsProps>(null);
+///
+///     useEffect(() {
+///       if (prevPropsRef.current != null) {
+///         print('old props: ${prevPropsRef.current}');
+///         print('new props: $props');
+///       }
+///
+///       prevPropsRef.current = props;
+///     });
+///
+///     return ((props.builder()
+///       ..addProps(getPropsToForward(props, onlyCopyDomProps: true))
+///       ..ref = ref)(props.children));
+///   },
+///   $LogPropsConfig, // ignore: undefined_identifier
+/// );
+///
+/// // ---------- Component Consumption ----------
+///
+/// void main() {
+///   final ref = createRef<Element>();
+///
+///   react_dom.render(
+///       (LogProps()
+///         ..builder = FancyButton
+///         ..className = 'btn btn-primary'
+///         ..ref = ref
+///         ..onClick = (_) {
+///           print(ref.current.outerHtml);
+///         }
+///       )(),
+///       querySelector('#idOfSomeNodeInTheDom')
+///   );
+///
+///   // You can still get a ref directly to the DOM button:
+///   final buttonNode = ref.current;
+/// }
+/// ```
+///
+/// Learn more: <https://reactjs.org/docs/forwarding-refs.html>.
+UiFactory<TProps> uiForwardRef<TProps extends bh.UiProps>(
+    dynamic Function(TProps props, dynamic ref) functionComponent, UiFactoryConfig<TProps> config) {
+  ArgumentError.checkNotNull(config, 'config');
+
+  // ignore: invalid_use_of_protected_member
+  var propsFactory = config.propsFactory;
+
+  // Get the display name from the inner function if possible so it doesn't become `_uiFunctionWrapper`
+  final displayName = config.displayName ?? getFunctionName(functionComponent);
+
+  dynamic _uiFunctionWrapper(dynamic props, dynamic ref) {
+    return functionComponent(propsFactory.jsMap(props), ref);
+  }
+
+  ReactJsComponentFactoryProxy factory;
+
+  // If a consumer uses `asForwardRefConfig` to generate the function component
+  // config, displayName could be `null`.
+  if (displayName != null && displayName.trim().isNotEmpty) {
```

Comment:
Why `displayName.trim().isNotEmpty`? 

This seems unnecessary

Reply:
Right?? It's weird. This may have just been covering up the symptom instead of the actual problem, so let me know if I should dig deeper. I added this though because if a consumer doesn't add a `displayName`, somehow it ends up ' '. I tried just `displayName.isEmpty`, and that didn't work. The result in the React devtools is (note the empty line that just says `forwardRef` - in pure ReactJS that would be 'Anonymous'):
![Screen Shot 2020-08-05 at 3 08 36 PM](https://user-images.githubusercontent.com/46691367/89469124-a7fb8400-d72d-11ea-9885-4f154406f3e8.png)

So yeah, it's weird and I spent a little time investigating before just adding this line.

Reply:
(I should definitely make the code comment more descriptive, at very least though)

Reply:
I would be curious how `' '` gets into displayName in the first place; I feel like that shouldn't happen

Reply:
Cool cool, I can look into that!

Reply:
Fixed here: d66db3f!

Reply:
I reverted that change - the issue was because anon functions in JS do have a name property, but it's an empty string. It seemed like an innocent change but to just have that `getFunctionName` function return `null` if it's empty, but that caused some breakages. react-dart also has a similar method that's identical, so it feels like we've built up a little bit of dependence on the behavior that null doesn't get returned. That makes me think this is still the simplest solution, but what do you think @greglittlefield-wf?

Reply:
That sounds good to me!